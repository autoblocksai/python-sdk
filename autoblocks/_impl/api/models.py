from dataclasses import dataclass
from typing import Dict
from typing import List
from typing import Optional

from autoblocks._impl.util import StrEnum


@dataclass
class Event:
    id: str
    trace_id: str
    message: str
    timestamp: str
    properties: Dict


@dataclass
class Trace:
    id: str
    events: List[Event]


@dataclass
class View:
    id: str
    name: str


@dataclass
class TracesResponse:
    next_cursor: Optional[str]
    traces: List[Trace]


@dataclass
class RelativeTimeFilter:
    type: str = "relative"
    seconds: Optional[float] = None
    minutes: Optional[float] = None
    hours: Optional[float] = None
    days: Optional[float] = None
    weeks: Optional[float] = None
    months: Optional[float] = None
    years: Optional[float] = None


@dataclass
class AbsoluteTimeFilter:
    start: str
    end: str
    type: str = "absolute"


class TraceFilterOperator(StrEnum):
    CONTAINS = "CONTAINS"
    NOT_CONTAINS = "NOT_CONTAINS"


class EventFilterOperator(StrEnum):
    CONTAINS = "CONTAINS"
    NOT_CONTAINS = "NOT_CONTAINS"
    EQUALS = "EQUALS"
    NOT_EQUALS = "NOT_EQUALS"
    LESS_THAN = "LESS_THAN"
    LESS_THAN_OR_EQUALS = "LESS_THAN_OR_EQUALS"
    GREATER_THAN = "GREATER_THAN"
    GREATER_THAN_OR_EQUALS = "GREATER_THAN_OR_EQUALS"


@dataclass
class EventFilter:
    key: str
    value: str
    operator: EventFilterOperator


@dataclass
class TraceFilter:
    operator: TraceFilterOperator
    event_filters: List[EventFilter]


class SystemEventFilterKey(StrEnum):
    MESSAGE = "SYSTEM:message"
    LABEL = "SYSTEM:label"
