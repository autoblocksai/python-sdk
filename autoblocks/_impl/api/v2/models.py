import functools
from typing import Any, Dict, List, Optional

import pydantic

from autoblocks._impl.prompts.placeholders import TemplatePlaceholder
from autoblocks._impl.prompts.placeholders import parse_placeholders_from_template


class PromptTemplate(pydantic.BaseModel):
    """An individual template for a prompt in the V2 API."""
    id: str
    template: str

    @functools.cached_property
    def placeholders(self) -> List[TemplatePlaceholder]:
        """Parse and return all placeholders in the template."""
        return parse_placeholders_from_template(self.template)


class PromptMajorVersion(pydantic.BaseModel):
    """A major version of a prompt in the V2 API."""
    major_version: str
    minor_versions: List[str]
    templates: List[PromptTemplate]
    tools_params: Optional[List[Dict[str, Any]]] = None
    params: Optional[Dict[str, Any]] = None


class Prompt(pydantic.BaseModel):
    """A prompt in the V2 API."""
    id: str
    app_id: str
    major_versions: List[PromptMajorVersion]

    def get_major_version(self, version: str) -> Optional[PromptMajorVersion]:
        """Get a specific major version of the prompt."""
        for major_version in self.major_versions:
            if major_version.major_version == version:
                return major_version
        return None 