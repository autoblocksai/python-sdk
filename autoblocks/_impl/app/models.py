"""App models for Autoblocks API."""

from pydantic import BaseModel
from pydantic import ConfigDict


class App(BaseModel):
    """App model representing an application."""

    id: str
    name: str
    slug: str

    model_config = ConfigDict(populate_by_name=True, extra="allow")
