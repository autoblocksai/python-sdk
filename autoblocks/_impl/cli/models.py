import pydantic

from autoblocks._impl.prompts.models import AutogeneratePromptsConfig


class FrozenModel(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(frozen=True)


class AutogenerateConfig(FrozenModel):
    prompts: AutogeneratePromptsConfig


class YamlConfig(FrozenModel):
    autogenerate: AutogenerateConfig
