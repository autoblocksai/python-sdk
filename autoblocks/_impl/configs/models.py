from dataclasses import dataclass
from typing import Any
from typing import List
from typing import Optional

try:
    import pydantic

    assert pydantic.__version__.startswith("2.")
except (ImportError, AssertionError):
    raise ImportError(
        "The Autoblocks config SDK requires pydantic version 2.x. "
        "You can install it with `pip install pydantic==2.*`."
    )


class FrozenModel(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(frozen=True)


class Property(FrozenModel):
    id: str
    value: Any


class RemoteConfigResponse(FrozenModel):
    id: str
    version: str
    properties: List[Property]


@dataclass
class RemoteConfig:
    id: str
    # only one of these should be specified
    major_version: Optional[str] = None  # should be a specific major version
    minor_version: Optional[str] = None  # should be a specific minor version or "latest"
    dangerously_use_undeployed_revision: Optional[str] = None  # can be a specific revision id or "latest"
