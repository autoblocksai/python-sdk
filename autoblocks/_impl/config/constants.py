API_ENDPOINT = "https://api.autoblocks.ai"
INGESTION_ENDPOINT = "https://ingest-event.autoblocks.ai"
AUTOBLOCKS_INGESTION_KEY = "AUTOBLOCKS_INGESTION_KEY"
AUTOBLOCKS_REPLAY_ID = "AUTOBLOCKS_REPLAY_ID"
