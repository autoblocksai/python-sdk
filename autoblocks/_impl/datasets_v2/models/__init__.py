"""Data models for Datasets V2 API."""

# Re-export all the models for easy access
from .conversation import (
    Conversation,
    ConversationMessage,
    ConversationTurn
)

from .dataset import (
    DatasetV2,
    DatasetListItemV2,
    DatasetSchemaV2,
    DatasetItemV2,
    DatasetItemsResponseV2,
    CreateDatasetV2Request,
    CreateDatasetItemsV2Request,
    UpdateItemV2Request
)

from .schema import (
    SchemaPropertyType,
    SchemaProperty,
    StringProperty,
    NumberProperty,
    BooleanProperty,
    ListOfStringsProperty,
    SelectProperty,
    MultiSelectProperty,
    ValidJSONProperty,
    ConversationProperty,
    create_schema_property
)

__all__ = [
    # Conversation models
    'Conversation',
    'ConversationMessage',
    'ConversationTurn',
    
    # Dataset models
    'DatasetV2',
    'DatasetListItemV2',
    'DatasetSchemaV2',
    'DatasetItemV2',
    'DatasetItemsResponseV2',
    'CreateDatasetV2Request',
    'CreateDatasetItemsV2Request',
    'UpdateItemV2Request',
    
    # Schema models
    'SchemaPropertyType',
    'SchemaProperty',
    'StringProperty',
    'NumberProperty',
    'BooleanProperty',
    'ListOfStringsProperty',
    'SelectProperty',
    'MultiSelectProperty',
    'ValidJSONProperty',
    'ConversationProperty',
    'create_schema_property',
] 