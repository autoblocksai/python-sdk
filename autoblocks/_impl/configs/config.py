import abc
import asyncio
import json
import logging
from datetime import timedelta
from typing import Any
from typing import Callable
from typing import Dict
from typing import Generic
from typing import Optional
from typing import TypeVar

from autoblocks._impl import global_state
from autoblocks._impl.config.constants import API_ENDPOINT
from autoblocks._impl.config.constants import REVISION_LATEST
from autoblocks._impl.configs.models import RemoteConfig
from autoblocks._impl.configs.models import RemoteConfigResponse
from autoblocks._impl.context_vars import RevisionType
from autoblocks._impl.context_vars import register_test_case_revision_usage
from autoblocks._impl.util import AnyTask
from autoblocks._impl.util import AutoblocksEnvVar
from autoblocks._impl.util import get_running_loop

log = logging.getLogger(__name__)

AutoblocksConfigValueType = TypeVar("AutoblocksConfigValueType")


def is_testing_context() -> bool:
    """
    Note that we check for the presence of the CLI environment
    variable and not the test case contextvars because the
    contextvars aren't set until run_test_suite is called,
    whereas a prompt manager might have already been imported
    and initialized by the time run_test_suite is called.
    """
    return bool(AutoblocksEnvVar.CLI_SERVER_ADDRESS.get())


def config_revisions_map() -> dict[str, str]:
    """
    The AUTOBLOCKS_CONFIG_REVISIONS environment variable is a JSON-stringified
    map of config IDs to revision IDs. This is set in CI test runs triggered
    from the UI.
    """
    if not is_testing_context():
        return {}

    config_revisions_raw = AutoblocksEnvVar.CONFIG_REVISIONS.get()
    if not config_revisions_raw:
        return {}
    return json.loads(config_revisions_raw)  # type: ignore


def make_request_url(config: RemoteConfig) -> str:
    config_id = config.id
    base = f"{API_ENDPOINT}/configs/{config_id}"
    if not isinstance(config, RemoteConfig) and (
        (config.major_version is None and config.minor_version is None)
        or config.dangerously_use_undeployed_revision is None
    ):
        raise ValueError(f"Invalid config type: {config}")

    if is_testing_context() and (revision_id := config_revisions_map().get(config_id)):
        return f"{base}/revisions/{revision_id}"

    if config.dangerously_use_undeployed_revision is not None:
        if config.dangerously_use_undeployed_revision == REVISION_LATEST:
            return f"{base}/revisions/{REVISION_LATEST}"
        else:
            return f"{base}/revisions/{config.dangerously_use_undeployed_revision}"

    if config.minor_version == REVISION_LATEST:
        return f"{base}/major/{config.major_version}/minor/{REVISION_LATEST}"

    return f"{base}/major/{config.major_version}/minor/{config.minor_version}"


def is_remote_config_refreshable(config: RemoteConfig) -> bool:
    """
    A config is refreshable if the user has specified that they want the latest deployed config
    or the latest undeployed config.
    """
    return (
        config.major_version is not None and config.minor_version == REVISION_LATEST
    ) or config.dangerously_use_undeployed_revision == REVISION_LATEST


async def get_remote_config(config: RemoteConfig, timeout: timedelta, api_key: str) -> RemoteConfigResponse:
    resp = await global_state.http_client().get(
        make_request_url(config=config),
        timeout=timeout.total_seconds(),
        headers={"Authorization": f"Bearer {api_key}"},
    )
    resp.raise_for_status()
    return RemoteConfigResponse.model_validate(resp.json())


class AutoblocksConfig(
    abc.ABC,
    Generic[AutoblocksConfigValueType],
):
    _value: AutoblocksConfigValueType
    _remote_config_id: Optional[str] = None
    _remote_config_revision_id: Optional[str] = None

    def __init__(self, value: AutoblocksConfigValueType) -> None:
        self._value = value

    async def _refresh_loop(
        self,
        config: RemoteConfig,
        refresh_interval: timedelta,
        refresh_timeout: timedelta,
        api_key: str,
        parser: Callable[[Dict[str, Any]], AutoblocksConfigValueType],
    ) -> None:
        """
        Refreshes the latest version of the config every `refresh_interval` seconds.

        This isn't exactly "every n seconds" but more "run and then wait n seconds",
        which will cause it to drift each run by the time it takes to run the refresh,
        but I think that is ok in this scenario.
        """
        while True:
            await asyncio.sleep(refresh_interval.total_seconds())

            try:
                await self._load_and_set_remote_config(
                    config=config, api_key=api_key, timeout=refresh_timeout, parser=parser
                )
            except Exception as err:
                log.warning(f"Failed to refresh config '{config.id}': {err}")

    async def _load_and_set_remote_config(
        self,
        config: RemoteConfig,
        api_key: str,
        timeout: timedelta,
        parser: Callable[[Dict[str, Any]], AutoblocksConfigValueType],
    ) -> None:
        """
        Loads the remote config from Autoblocks and sets the value of this config
        based on the result.

        If the parser is specified, the value will be parsed using the parser.
        """
        remote_config = await get_remote_config(config, timeout=timeout, api_key=api_key)
        try:
            obj = {}
            for prop in remote_config.properties:
                obj[prop.id] = prop.value
            parsed = parser(obj)
            if parsed is not None:
                self._value = parsed
                self._remote_config_id = remote_config.id
                self._remote_config_revision_id = remote_config.revision_id
        except Exception as err:
            log.error(f"Failed to parse config '{config.id}': {err}", exc_info=True)
            raise err

    def _activate_from_remote_unsafe(
        self,
        config: RemoteConfig,
        refresh_interval: timedelta,
        refresh_timeout: timedelta,
        activate_timeout: timedelta,
        parser: Callable[[Dict[str, Any]], AutoblocksConfigValueType],
        api_key: Optional[str] = None,
    ) -> None:
        api_key = api_key or AutoblocksEnvVar.API_KEY.get()
        if not api_key:
            raise ValueError(
                f"You must either pass in the API key via 'api_key' or "
                f"set the {AutoblocksEnvVar.API_KEY} environment variable."
            )

        global_state.init()
        task: AnyTask
        if running_loop := get_running_loop():
            # If we're already in a running loop, execute the task on that loop
            task = running_loop.create_task(
                self._load_and_set_remote_config(
                    config=config, api_key=api_key, timeout=activate_timeout, parser=parser
                ),
            )
        else:
            # Otherwise, send the task to our background loop
            task = asyncio.run_coroutine_threadsafe(
                self._load_and_set_remote_config(
                    config=config, api_key=api_key, timeout=activate_timeout, parser=parser
                ),
                global_state.event_loop(),
            )

        # Wait for config to be initialized
        task.result()

        # we do not refresh the config if we are in a testing context
        if is_remote_config_refreshable(config) and not is_testing_context():
            if running_loop := get_running_loop():
                running_loop.create_task(
                    self._refresh_loop(
                        config=config,
                        api_key=api_key,
                        refresh_interval=refresh_interval,
                        refresh_timeout=refresh_timeout,
                        parser=parser,
                    )
                )
            else:
                asyncio.run_coroutine_threadsafe(
                    self._refresh_loop(
                        config=config,
                        api_key=api_key,
                        refresh_interval=refresh_interval,
                        refresh_timeout=refresh_timeout,
                        parser=parser,
                    ),
                    global_state.event_loop(),
                )

    def activate_from_remote(
        self,
        config: RemoteConfig,
        parser: Callable[[Dict[str, Any]], AutoblocksConfigValueType],
        api_key: Optional[str] = None,
        refresh_interval: timedelta = timedelta(seconds=10),
        refresh_timeout: timedelta = timedelta(seconds=30),
        activate_timeout: timedelta = timedelta(seconds=30),
    ) -> None:
        """
        Activate a remote config from Autoblocks and optionally refresh it every `refresh_interval` seconds.
        """
        try:
            log.info("Activating remote config...")
            self._activate_from_remote_unsafe(
                config=config,
                api_key=api_key,
                refresh_interval=refresh_interval,
                refresh_timeout=refresh_timeout,
                activate_timeout=activate_timeout,
                parser=parser,
            )
        except Exception as err:
            log.error(f"Failed to activate remote config '{config.id}': {err}")

    @property
    def value(self) -> AutoblocksConfigValueType:
        if is_testing_context() and self._remote_config_id and self._remote_config_revision_id:
            register_test_case_revision_usage(
                entity_id=self._remote_config_id,
                entity_type=RevisionType.CONFIG,
                revision_id=self._remote_config_revision_id,
            )
        return self._value
