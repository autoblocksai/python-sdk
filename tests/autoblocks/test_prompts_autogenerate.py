import os
from unittest import mock

from autoblocks._impl.config.constants import API_ENDPOINT
from autoblocks._impl.prompts.autogenerate import infer_type
from autoblocks._impl.prompts.autogenerate import to_snake_case
from autoblocks._impl.prompts.autogenerate import to_title_case
from autoblocks._impl.prompts.autogenerate import write_generated_code_for_config
from autoblocks._impl.prompts.models import AutogeneratePromptConfig
from autoblocks._impl.prompts.models import AutogeneratePromptsConfig
from autoblocks._impl.prompts.placeholders import TemplatePlaceholder
from autoblocks._impl.prompts.placeholders import parse_placeholders_from_template

# Show full diff in unittest
# import unittest
# unittest.util._MAX_LENGTH = 2000


def test_to_title_case():
    assert to_title_case("hello") == "Hello"
    assert to_title_case("helloWorld") == "HelloWorld"
    assert to_title_case("hello world") == "HelloWorld"
    assert to_title_case("Hello World") == "HelloWorld"
    assert to_title_case("hello_world") == "HelloWorld"
    assert to_title_case("hello-world") == "HelloWorld"
    assert to_title_case("hello/world") == "HelloWorld"
    assert to_title_case("hello123world") == "Hello123world"
    assert to_title_case("hello-123-world") == "Hello123World"
    assert to_title_case("hello-123") == "Hello123"
    assert to_title_case("123hello-world") == "HelloWorld"


def test_to_snake_case():
    assert to_snake_case("hello") == "hello"
    assert to_snake_case("helloWorld") == "hello_world"
    assert to_snake_case("hello world") == "hello_world"
    assert to_snake_case("hello    world") == "hello_world"
    assert to_snake_case("Hello World") == "hello_world"
    assert to_snake_case("hello_world") == "hello_world"
    assert to_snake_case("hello___world") == "hello_world"
    assert to_snake_case("hello-world") == "hello_world"
    assert to_snake_case("hello---world") == "hello_world"
    assert to_snake_case("hello/world") == "hello_world"
    assert to_snake_case("Hello/World") == "hello_world"
    assert to_snake_case("hello123world") == "hello123world"
    assert to_snake_case("hello-123-world") == "hello_123_world"
    assert to_snake_case("hello-123") == "hello_123"
    assert to_snake_case("123hello-world") == "hello_world"


def test_parse_placeholders_from_template():
    assert parse_placeholders_from_template("Hello") == []
    assert parse_placeholders_from_template("Hello, {{ name }}!") == [
        TemplatePlaceholder(
            is_escaped=False,
            name="name",
        ),
    ]
    assert parse_placeholders_from_template("Hello, {{ name }}! My name is {{ name }}.") == [
        TemplatePlaceholder(
            is_escaped=False,
            name="name",
        ),
    ]
    assert parse_placeholders_from_template("{{ c }} {{ b }} {{a}}") == [
        TemplatePlaceholder(
            is_escaped=False,
            name="a",
        ),
        TemplatePlaceholder(
            is_escaped=False,
            name="b",
        ),
        TemplatePlaceholder(
            is_escaped=False,
            name="c",
        ),
    ]
    assert parse_placeholders_from_template("{{ param with spaces }} {{ param-with-hyphens }}") == [
        TemplatePlaceholder(
            is_escaped=False,
            name="param-with-hyphens",
        ),
    ]
    assert parse_placeholders_from_template("I am \\{{ escaped }} and I am {{ not }} escaped") == [
        TemplatePlaceholder(
            is_escaped=True,
            name="escaped",
        ),
        TemplatePlaceholder(
            is_escaped=False,
            name="not",
        ),
    ]
    assert parse_placeholders_from_template("\\{{ escaped }} \\{{ escaped }} {{ escaped }}") == [
        TemplatePlaceholder(
            is_escaped=False,
            name="escaped",
        ),
        TemplatePlaceholder(
            is_escaped=True,
            name="escaped",
        ),
    ]
    assert (
        parse_placeholders_from_template(
            """Please respond in the format:

{{
  "x": {{
    "y": 1
  }}
}}

Thanks!
"""
        )
        == []
    )


def test_infer_type():
    assert infer_type("") == "str"
    assert infer_type(0) == "Union[float, int]"
    assert infer_type(0.0) == "Union[float, int]"
    assert infer_type(True) == "bool"
    assert infer_type(False) == "bool"
    assert infer_type([""]) == "list[str]"
    assert infer_type([]) is None
    assert infer_type({}) == "Dict[str, Any]"
    assert infer_type({"a": 1}) == "Dict[str, Any]"
    assert infer_type(None) is None


@mock.patch.dict(
    os.environ,
    {
        "AUTOBLOCKS_API_KEY": "mock-api-key",
    },
)
def test_write(httpx_mock, snapshot):
    config = AutogeneratePromptsConfig(
        outfile="test.py",
        prompts=[
            AutogeneratePromptConfig(
                id="prompt-a",
                major_version="1",
            ),
            AutogeneratePromptConfig(
                id="prompt-b",
                # Test that it works with numbers
                major_version=1,  # type: ignore
            ),
            AutogeneratePromptConfig(
                id="prompt-c",
                dangerously_use_undeployed_revision="latest",
            ),
            AutogeneratePromptConfig(
                id="prompt-d",
                dangerously_use_undeployed_revision="mock-specific-revision-id",
            ),
        ],
    )

    httpx_mock.add_response(
        url=f"{API_ENDPOINT}/prompts/prompt-a/major/1/minor/latest",
        method="GET",
        status_code=200,
        json={
            "id": "prompt-a",
            "params": {
                "params": {
                    "frequencyPenalty": 0,
                    "maxTokens": 256,
                    "model": "gpt-4",
                    "presencePenalty": 0.3,
                    "stopSequences": [],
                    "temperature": 0.7,
                    "topP": 1,
                    "seed": 4096,
                    "responseFormat": {"type": "json_object"},
                },
                "version": "1.0",
            },
            "templates": [
                {
                    "id": "template-a",
                    "template": "Hello, {{ name }}! The weather is {{ weather }} today.",
                },
                {
                    "id": "template-b",
                    "template": "Hello, {{ name-1 }}! My name is {{ name-2 }}. "
                    "Escaped \{{escaped}} placeholders should be ignored.",
                },
            ],
            "toolsParams": [],
            "version": "1.0",
            "revisionId": "mock-revision-id",
        },
        match_headers={"Authorization": "Bearer mock-api-key"},
    )

    httpx_mock.add_response(
        url=f"{API_ENDPOINT}/prompts/prompt-b/major/1/minor/latest",
        method="GET",
        status_code=200,
        json={
            "id": "prompt-b",
            "params": {
                "params": {
                    "frequencyPenalty": 0,
                    "maxTokens": 256,
                    "model": "gpt-4",
                    "presencePenalty": -0.3,
                    "stopSequences": [],
                    "temperature": 0.7,
                    "topP": 1,
                    "seed": 4096,
                    "responseFormat": {"type": "json_object"},
                },
            },
            "templates": [
                {
                    "id": "template-a",
                    "template": "Hello, {{ name }}! The weather is {{ weather }} today.",
                },
            ],
            "toolsParams": [{"name": "tool-b", "params": ["description"]}],
            "version": "3.1",
            "revisionId": "mock-revision-id",
        },
        match_headers={"Authorization": "Bearer mock-api-key"},
    )

    httpx_mock.add_response(
        url=f"{API_ENDPOINT}/prompts/prompt-c/major/undeployed/minor/latest",
        method="GET",
        status_code=200,
        json={
            "id": "prompt-c",
            "params": None,
            "templates": [
                {
                    "id": "template-a",
                    "template": "Hello, {{ name }}! The weather is {{ weather }} today.",
                },
            ],
            "toolsParams": [],
            "version": "revision:mock-revision-id",
            "revisionId": "mock-revision-id",
        },
        match_headers={"Authorization": "Bearer mock-api-key"},
    )

    httpx_mock.add_response(
        url=f"{API_ENDPOINT}/prompts/prompt-d/major/undeployed/minor/mock-specific-revision-id",
        method="GET",
        status_code=200,
        json={
            "id": "prompt-c",
            "params": None,
            "templates": [
                {
                    "id": "template-a",
                    "template": "Hello, {{ name }}! The weather is {{ weather }} today.",
                },
            ],
            "toolsParams": [],
            "version": "revision:mock-specific-revision-id",
            "revisionId": "mock-specific-revision-id",
        },
        match_headers={"Authorization": "Bearer mock-api-key"},
    )

    m = mock.mock_open()

    with mock.patch("autoblocks._impl.prompts.autogenerate.open", m):
        write_generated_code_for_config(config)

    m.assert_called_once_with("test.py", "w")
    m().write.assert_called_once()
    write_called_with = m().write.call_args[0][0]
    assert write_called_with == snapshot
