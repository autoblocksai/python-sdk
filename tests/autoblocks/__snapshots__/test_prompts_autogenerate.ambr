# serializer version: 1
# name: test_write
  '''
  ############################################################################
  # This file was generated automatically by Autoblocks. Do not edit directly.
  ############################################################################
  
  from typing import Any  # noqa: F401
  from typing import Dict  # noqa: F401
  from typing import Union  # noqa: F401
  
  import pydantic  # noqa: F401
  
  from autoblocks.prompts.context import PromptExecutionContext
  from autoblocks.prompts.manager import AutoblocksPromptManager
  from autoblocks.prompts.models import FrozenModel
  from autoblocks.prompts.renderer import TemplateRenderer
  from autoblocks.prompts.renderer import ToolRenderer
  
  
  class PromptAParams(FrozenModel):
      frequency_penalty: Union[float, int] = pydantic.Field(..., alias="frequencyPenalty")
      max_completion_tokens: Union[float, int] = pydantic.Field(..., alias="maxCompletionTokens")
      model: str = pydantic.Field(..., alias="model")
      presence_penalty: Union[float, int] = pydantic.Field(..., alias="presencePenalty")
      temperature: Union[float, int] = pydantic.Field(..., alias="temperature")
      top_p: Union[float, int] = pydantic.Field(..., alias="topP")
      seed: Union[float, int] = pydantic.Field(..., alias="seed")
      response_format: Dict[str, Any] = pydantic.Field(..., alias="responseFormat")
  
  
  class PromptATemplateRenderer(TemplateRenderer):
      __name_mapper__ = {
          "name": "name",
          "name-1": "name_1",
          "name-2": "name_2",
          "weather": "weather",
      }
  
      def template_a(
          self,
          *,
          name: str,
          weather: str,
      ) -> str:
          return self._render(
              "template-a",
              name=name,
              weather=weather,
          )
  
      def template_b(
          self,
          *,
          name_1: str,
          name_2: str,
      ) -> str:
          return self._render(
              "template-b",
              name_1=name_1,
              name_2=name_2,
          )
  
  
  class PromptAToolRenderer(ToolRenderer):
      __name_mapper__ = {}
  
  
  
  class PromptAExecutionContext(
      PromptExecutionContext[
          PromptAParams,
          PromptATemplateRenderer,
          PromptAToolRenderer,
      ],
  ):
      __params_class__ = PromptAParams
      __template_renderer_class__ = PromptATemplateRenderer
      __tool_renderer_class__ = PromptAToolRenderer
  
  
  class PromptAPromptManager(
      AutoblocksPromptManager[PromptAExecutionContext],
  ):
      __prompt_id__ = "prompt-a"
      __prompt_major_version__ = "1"
      __execution_context_class__ = PromptAExecutionContext
  
  
  class PromptBParams(FrozenModel):
      frequency_penalty: Union[float, int] = pydantic.Field(..., alias="frequencyPenalty")
      max_completion_tokens: Union[float, int] = pydantic.Field(..., alias="maxCompletionTokens")
      model: str = pydantic.Field(..., alias="model")
      presence_penalty: Union[float, int] = pydantic.Field(..., alias="presencePenalty")
      temperature: Union[float, int] = pydantic.Field(..., alias="temperature")
      top_p: Union[float, int] = pydantic.Field(..., alias="topP")
      seed: Union[float, int] = pydantic.Field(..., alias="seed")
      response_format: Dict[str, Any] = pydantic.Field(..., alias="responseFormat")
  
  
  class PromptBTemplateRenderer(TemplateRenderer):
      __name_mapper__ = {
          "name": "name",
          "weather": "weather",
      }
  
      def template_a(
          self,
          *,
          name: str,
          weather: str,
      ) -> str:
          return self._render(
              "template-a",
              name=name,
              weather=weather,
          )
  
  
  class PromptBToolRenderer(ToolRenderer):
      __name_mapper__ = {
          "description": "description",
      }
  
      def tool_b(
          self,
          *,
          description: str,
      ) -> Dict[str, Any]:
          return self._render(
              "tool-b",
              description=description,
          )
  
  
  class PromptBExecutionContext(
      PromptExecutionContext[
          PromptBParams,
          PromptBTemplateRenderer,
          PromptBToolRenderer,
      ],
  ):
      __params_class__ = PromptBParams
      __template_renderer_class__ = PromptBTemplateRenderer
      __tool_renderer_class__ = PromptBToolRenderer
  
  
  class PromptBPromptManager(
      AutoblocksPromptManager[PromptBExecutionContext],
  ):
      __prompt_id__ = "prompt-b"
      __prompt_major_version__ = "1"
      __execution_context_class__ = PromptBExecutionContext
  
  
  class PromptCParams(FrozenModel):
      pass
  
  
  class PromptCTemplateRenderer(TemplateRenderer):
      __name_mapper__ = {
          "name": "name",
          "weather": "weather",
      }
  
      def template_a(
          self,
          *,
          name: str,
          weather: str,
      ) -> str:
          return self._render(
              "template-a",
              name=name,
              weather=weather,
          )
  
  
  class PromptCToolRenderer(ToolRenderer):
      __name_mapper__ = {}
  
  
  
  class PromptCExecutionContext(
      PromptExecutionContext[
          PromptCParams,
          PromptCTemplateRenderer,
          PromptCToolRenderer,
      ],
  ):
      __params_class__ = PromptCParams
      __template_renderer_class__ = PromptCTemplateRenderer
      __tool_renderer_class__ = PromptCToolRenderer
  
  
  class PromptCPromptManager(
      AutoblocksPromptManager[PromptCExecutionContext],
  ):
      __prompt_id__ = "prompt-c"
      __prompt_major_version__ = "undeployed"
      __execution_context_class__ = PromptCExecutionContext
  
  
  class PromptDParams(FrozenModel):
      pass
  
  
  class PromptDTemplateRenderer(TemplateRenderer):
      __name_mapper__ = {
          "name": "name",
          "weather": "weather",
      }
  
      def template_a(
          self,
          *,
          name: str,
          weather: str,
      ) -> str:
          return self._render(
              "template-a",
              name=name,
              weather=weather,
          )
  
  
  class PromptDToolRenderer(ToolRenderer):
      __name_mapper__ = {}
  
  
  
  class PromptDExecutionContext(
      PromptExecutionContext[
          PromptDParams,
          PromptDTemplateRenderer,
          PromptDToolRenderer,
      ],
  ):
      __params_class__ = PromptDParams
      __template_renderer_class__ = PromptDTemplateRenderer
      __tool_renderer_class__ = PromptDToolRenderer
  
  
  class PromptDPromptManager(
      AutoblocksPromptManager[PromptDExecutionContext],
  ):
      __prompt_id__ = "prompt-d"
      __prompt_major_version__ = "undeployed"
      __execution_context_class__ = PromptDExecutionContext
  
  '''
# ---
